### PR TITLE
mat64: add overlap detection for blas64 returning types

### PR DESCRIPTION
**Pull request by kortschak**
This provides the necessary tools for adding shadowing detection. There is currently no testing for the symmetric or triangular matrix overlaps methods, but they are substantively the same as the general matrix case and it is not yet clear how views of those types will arise - clients can do it and I guess we might at some stage add ViewSym and ViewTri or some near equivalent.

The test is conservative in general (mismatched stride makes a yes if the matrix data slices overlap at all - this should never happen), but more so with the symmetric and triangular cases because of ease of coding.

Updates #200.

@btracey Please take a look.

### CODE REVIEW

**Comment by kortschak** [comment]
I've written in the specification for the aliasing behaviour, which is slightly more safe than the version written in the original issue - the docs include pointer equality (which was not originally included), mainly because it's easy. Please ensure I've been honest here - I may have missed some cases of that test since I only decided to formalise it half way through the series (I went back and amended some commits - hopefully I got all of them).

**Comment by btracey** [comment]
High level discussion before any detail -- what behavior do we want? It seems to me like we want 1) To enable efficient computation, both in memory and clock cycles 2) To support and enforce correct code 3) To assist ease and readability

This PR has the mentality of "allocate defensively where possible, and permissive otherwise". Instead, we could have the mentality of "Don't allow overlap, unless we can prove that it's okay". We could say something along the lines of "Sharing data between src and dst is not allowed, unless it can be proven to be okay. As an exception to this rule, we allow sharing in the case of pointer equality". We can then comment along the lines of where you have now, that we can prove it is okay when the types are the same, the strides are the same, and there is no overlap.

I feel like "defensive panicking" may be better than "defensive allocating". You say that the strides should never be unequal, but it's really easy to construct
````
data := make([]float64, 64)
a := mat64.NewDense(8, 8, data)
b := mat64.NewDense(4, 16, data)
````
Or worse
````
c := mat64.NewDense(8, 4, data[:32])
````
If the user writes
````
c.Mul(a, d)
````
Do we really want to allocate defensively, or would we rather instead panic because of disallowed aliasing? It seems to me like allowing aliasing leads to code that can be very hard to reason about. Instead, we can just disallow it, provide the special case of pointer equality (because it's really useful), and have some decent aliasing detection (as in the case of equal stride but no overlap). If a user really wants to do crazy stuff with aliased src and dst matrices, we have the blas64 and lapack64 libraries for them to call directly.

**Comment by btracey** [comment]
To rephrase my comment above, we could have something like

````
Element Aliasing

Most methods in mat64 modify receiver data. It is forbidden for the modified data region of the receiver to overlap the used data area of the input arguments. The exception to this rule is if the method receiver is equal to one of the input arguments, as in the a.Pow(a, 6) call above. When data is shared between the receiver and input, mat64 will attempt to detect if the read and write regions overlap, but if it cannot prove no overlap occurs, the method will panic. At present, the following cases will not panic 
- Data doesn't overlap
- Receiver pointer equal to input pointer
- Other stuff you have (types match, strides equal, etc.)
````

We can even throw in the warning from the memory model: "If you need to read this section to understand the behavior of your program, you are being clever. Don't be clever. If you must be clever, blas64 and lapack64 may be used to call the behavior directly."

**Comment by kortschak** [comment]
The philosophical implications are pretty deep and we will be stuck with it when it's in and we freeze the API, so I want to be sure that we get this right. I think the decision about this should happen on gonum-dev. Two opinions is not enough - I'm not wedded to the PR as is, but one of the two approaches (or something better) does need to happen.

The changes necessary other than doc should be simple (your words are pretty close to what it should be if that approach is taken - I like the suggestion about cribbing from MM), though I need to check all the calls - it really should just be a replacement in the overlaps methods of "return true" with "panic(badRegion)" and a conditional panic at the tail, where badRegion has a meaningful text.

**Comment by btracey** [comment]
Agreed on all fronts.

**Comment by kortschak** [comment]
I've made the minimal change as an experiment and all tests pass except for TestSolveVec which I am trying to figure out (there are some additions to list_test that should come from this).

It seems that someone ;) is doing something clever by coercing a *Vector into a *Dense. This results in an exact match between the receiver and the parameter, but pointer difference.

**Comment by btracey** [comment]
Ohhhh, that is tricky. There are two solutions. 1) Duplicate the code 2) Test for pointer equality. If pointer equality holds, then call Dense.Solve with the same pointer, if it doesn't hold, then call Dense.Solve is it is.

**Comment by kortschak** [comment]
If I add the following code for the `off == 0` case, all tests pass, but that breaks the invariant that we don't allow operations to modify the parameter.

```
		if m.mat.Rows == a.Rows && m.mat.Cols == a.Cols && m.mat.Stride == a.Stride {
			return false
		}
```

**Comment by btracey** [comment]
You mean it breaks the invariant because it changes the data even if the pointers are unequal? We should probably be discouraging that kind of cleverness, especially since there's an easy workaround that makes the behavior explicit (what I mentioned with testing for equality).

**Comment by kortschak** [comment]
There is not opportunity to test for pointer equality by the time overlaps has the data, the blas.General (or other) type has already been extracted from the RawMatrixer.

We can fix the problem earlier though. This fixes it:

```
func (v *Vector) SolveVec(a Matrix, b *Vector) error {
	_, c := a.Dims()
	// The Solve implementation is non-trivial, so rather than duplicate the code,
	// instead recast the Vectors as Dense and call the matrix code.
	v.reuseAs(c)
	m := vecAsDense(v)
	bm := m
	if v != b {
		bm = vecAsDense(b)
	}
	return m.Solve(a, bm)
}
```

**Comment by btracey** [comment]
Yes, sorry, I should have been more clear. That is what I was thinking. I was actually thinking 
````
v.reuseAs(c)
m := vecAsDense(v)
if v == b {
   return m.Solve(a, m)
}
bm := vecAsDense(b)
return m.Solve(a, bm)
````
Which is what I meant by "if pointer equality holds, call Solve with the same pointer". Your code is better anyway.

**Comment by kortschak** [comment]
Also, I need to add check code for *Vector in the first commit here.

**Comment by kortschak** [comment]
I'll add that to the solve routines commit either way. If we go with permissive, it save an allocation.

**Comment by kortschak** [comment]
Approach discussion at the gonum-dev thread, please.

https://groups.google.com/d/topic/gonum-dev/hQiEE0K8Kac/discussion

**Comment by kortschak** [comment]
Counter proposal here: https://github.com/gonum/matrix/tree/shadow-strict

**Comment by kortschak** [comment]
A week has gone past with no input in either direction, so with off-line discussions, I'm changing the proposal to be strict rather than permissive.

@btracey PTAL

**Comment by kortschak** [comment]
@btracey Gentle ping. I'd like to get this out of the way since it touches many scattered lines.

File: mat64/dense_arithmetic.go [hunk 1]
```diff
@@ -30,17 +30,17 @@ func (m *Dense) Add(a, b Matrix) {
...
-		m, restore = m.isolatedWorkspace(bMat)
-		defer restore()
-	}
 
-	if a, ok := a.(RawMatrixer); ok {
-		if b, ok := b.(RawMatrixer); ok {
-			amat, bmat := a.RawMatrix(), b.RawMatrix()
+	if arm, ok := a.(RawMatrixer); ok {
+		if brm, ok := b.(RawMatrixer); ok {
+			amat, bmat := arm.RawMatrix(), brm.RawMatrix()
```
Comment (btracey) [inline mat64/dense_arithmetic.go]: Given the algorithm, we don't need to isolate the workspace, just check overlap, correct?

File: mat64/dense_arithmetic.go [hunk 2]
```diff
@@ -85,17 +93,17 @@ func (m *Dense) Sub(a, b Matrix) {
...
```
Comment (btracey) [inline mat64/dense_arithmetic.go]: Same as add.

File: mat64/dense_arithmetic.go [hunk 3]
```diff
@@ -141,17 +157,17 @@ func (m *Dense) MulElem(a, b Matrix) {
...
```
Comment (btracey) [inline mat64/dense_arithmetic.go]: and here as well.

File: mat64/dense_arithmetic.go [hunk 4]
```diff
@@ -251,13 +283,28 @@ func (m *Dense) Inverse(a Matrix) error {
...
+				tmp := getWorkspace(r, c, false)
+				tmp.Copy(a)
+				m.Copy(tmp)
+				putWorkspace(tmp)
+				break
+			}
+			m.Copy(a)
+		}
+	default:
+		if m != aMat {
```
Comment (btracey) [inline mat64/dense_arithmetic.go]: If aMat is not a RawMatrixer, then it is impossible for m to equal aMat, isn't it? Thus it should just be m.Copy(a)?

File: mat64/dense_arithmetic.go [hunk 5]
```diff
@@ -655,8 +711,13 @@ func (m *Dense) Scale(f float64, a Matrix) {
 	m.reuseAs(ar, ac)
 
 	aMat, aTrans := untranspose(a)
-	if a, ok := aMat.(RawMatrixer); ok {
-		amat := a.RawMatrix()
+	if rm, ok := aMat.(RawMatrixer); ok {
```
Comment (btracey) [inline mat64/dense_arithmetic.go]: Here is the same as the add case.

File: mat64/dense_arithmetic.go [hunk 6]
```diff
@@ -703,8 +761,13 @@ func (m *Dense) Apply(fn func(r, c int, v float64) float64, a Matrix) {
...
+		amat := rm.RawMatrix()
```
Comment (btracey) [inline mat64/dense_arithmetic.go]: And here for the Add case as well I think.

File: mat64/doc.go [hunk 7]
```diff
@@ -112,7 +112,33 @@
...
-// detect arbitrary overlap, and so care should be taken when using matrix views.
+// though in many cases this will cause an allocation (see Element Aliasing).
 // An exception to this rule is Copy, which does not allow a.Copy(a.T()).
+//
+// Element Aliasing
+//
+// Most methods in mat64 modify receiver data. It is forbidden for the modified
+// data region of the receiver to overlap the used data area of the input
+// arguments. The exception to this rule is when the method receiver is equal to one
+// of the input arguments, as in the a.Pow(a, 6) call above.
```
Comment (btracey) [inline mat64/doc.go]: "Equal to one of the input arguments or its transpose"? Are we allowing a.Mul(a.T(), b)?

File: mat64/dense_arithmetic.go (hunk 1 above)
Comment (kortschak) [inline mat64/dense_arithmetic.go]: The idiom I've used here is explained in https://github.com/gonum/matrix/pull/263/files#diff-928774397024853172d191bc227c4939R23. It is somewhat misleading and I apologise for that; if given enough of an argument I would change it, but I would rather not.

In short, if checkOverlap does not return false, it panics, so the check is not one that will result in an isolation. The reason for doing it this way is that it keeps the check and the pointer comparison next to each other, which I think balances out the C-ism.

File: mat64/offset_appengine.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+	a0 := reflect.ValueOf(a).Index(0).UnsafeAddr()
+	b0 := reflect.ValueOf(b).Index(0).UnsafeAddr()
+
+	if a0 == b0 {
+		return 0
+	}
+	if a0 < b0 {
+		return int((b0 - a0) / sizeOfFloat64)
+	}
+	return -int((a0 - b0) / sizeOfFloat64)
```
Comment (btracey) [inline mat64/offset_appengine.go]: This doesn't look right. Hasn't the negative sign been added twice?

File: mat64/dense_arithmetic.go (hunk 4 above)
Comment (kortschak) [inline mat64/dense_arithmetic.go]: No, it can be a type conversion.

File: mat64/doc.go (hunk 7 above)
Comment (kortschak) [inline mat64/doc.go]: I'll add that, though that wording is not precise enough. I have to think about that.

File: mat64/offset_appengine.go (hunk 8 above)
Comment (kortschak) [inline mat64/offset_appengine.go]: a0 and b0 are uintptr, so I have to do arithmetic in natural space and then convert the sign.

File: mat64/shadow.go [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+	mismatchedStrides = "mat64: bad region: different strides"
+)
+
+// checkOverlap returns false if the receiver does not overlap data elements
+// referenced by the parameter and panics otherwise.
+//
+// checkOverlap methods return a boolean to allow the check call to be added to a
+// boolean expression, making use of short-circuit operators.
+
+func (m *Dense) checkOverlap(a blas64.General) bool {
```
Comment (btracey) [inline mat64/shadow.go]: It's easy to test for all Raw types, isn't it? If this takes in Matrix, then a type switch can be used to extract the data slice from any blas type. If there is any overlap and the types are different, we panic. If both are RawMatrixers, then we can do the extended testing.

File: mat64/offset_appengine.go (hunk 8 above)
Comment (btracey) [inline mat64/offset_appengine.go]: Oh, sorry, I was thinking this returned the absolute value, but it does not.

File: mat64/dense_arithmetic.go (hunk 1 above)
Comment (btracey) [inline mat64/dense_arithmetic.go]: The idiom is nice, but my point is that if m == a, the algorithm without allocating an isolated workspace.

File: mat64/dense_arithmetic.go (hunk 4 above)
Comment (btracey) [inline mat64/dense_arithmetic.go]: Ah, tricky.

File: mat64/shadow.go (hunk 9 above)
Comment (btracey) [inline mat64/shadow.go]: It seems like we could make it simple with something like

````
func (m *Dense) checkOverlap(a Matrix) bool {
    if m == a {
         return false
    }
     // Type switch to extract the data. If a is none of the Raw types, return false since we can't check.
    // If offset is in the right range, and the types are different, panic
    // If a is a RawMatrixer, do the fancier checks
}
````
This way, we can do something like
````
func (m *Dense) Method(a Matrix) {
    // Check size
    aMat, _ := untranspose(a)
    m.checkOverlap(aMat)
}
````
Thus we can check all of the necessary overlap at the top of the function once. We could even remove the boolean.

Comment (kortschak) [inline mat64/shadow.go]: This adds an extra assertion in many places. I considered it, but I don't like it.

File: mat64/dense_arithmetic.go (hunk 1 above)
Comment (kortschak) [inline mat64/dense_arithmetic.go]: If that's the case it was incorrect before this.

File: mat64/dense_arithmetic.go (hunk 5 above)
Comment (kortschak) [inline mat64/dense_arithmetic.go]: Tested. This one is required.

File: mat64/shadow.go (hunk 9 above)
Comment (btracey) [inline mat64/shadow.go]: I see. So the idea is that it's the users problem if the types are different and there is incorrect overlap. That's probably reasonable since you have to explicitly construct two different matrices with the same underlying data. The one exception is RowView and ColView of Dense. 
````
v := m.ColView(1)
v.SolveVec(m, v)
````
This is an easy case to construct, but will be incorrect. I'm not exactly sure what the correct conclusion is.

File: mat64/dense_arithmetic.go (hunk 1 above)
Comment (kortschak) [inline mat64/dense_arithmetic.go]: Done.

File: mat64/dense_arithmetic.go (hunk 2 above)
Comment (kortschak) [inline mat64/dense_arithmetic.go]: Done.

File: mat64/dense_arithmetic.go (hunk 3 above)
Comment (kortschak) [inline mat64/dense_arithmetic.go]: Done.

File: mat64/dense_arithmetic.go (hunk 6 above)
Comment (kortschak) [inline mat64/dense_arithmetic.go]: Also required.

File: mat64/shadow.go (hunk 9 above)
Comment (kortschak) [inline mat64/shadow.go]: This is why the documentation says what it says. Maybe we should special-case vectors?

**Comment by kortschak** [comment]
PTAL

File: mat64/shadow.go (hunk 9 above)
Comment (btracey) [inline mat64/shadow.go]: I'm just making sure I understand the intended behavior and its reasons (SGTM, just internalizing). But yea, possibly vectors should be special cased. Having ColView seems desirable.

Comment (kortschak) [inline mat64/shadow.go]: Thinking about this more. We must special-case vectors. The spirit of the documentation is that if you do things within the mat64 API we will protect you from yourself but if you do the equivalent of unsafe (which is why I originally called "RawMatrix" "UnsafeMatrix") then you are on your own. This is clearly within the API, and I would think possibly even something we should be permissive with, though again I'm not strongly bound to that position.

File: mat64/shadow.go [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+	}
+
+	if mat.Inc == 1 || off&mat.Inc == 0 {
+		panic(regionOverlap)
+	}
+	return false
+}
+
+// rectanglesOverlap returns whether the strided rectangles a and b overlap
+// when b is offset off elements after a but has at least one element before
```
Comment (btracey) [inline mat64/shadow.go]: Could you put "off" in quotes or rephrase? The documentation is really confusing when one tries to read "off" as a word instead of a variable name.

File: mat64/shadow.go (hunk 9 above)
Comment (btracey) [inline mat64/shadow.go]: I don't think we should be permissive with vectors in the same way we are permissive in pointer equality. At best, it's confusing code. I do agree with your interpretation of the spirit of the rules (protect within the API), and thus the need to add a check for overlap with vectors.

File: mat64/shadow.go [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+// checkOverlap methods return a boolean to allow the check call to be added to a
+// boolean expression, making use of short-circuit operators.
+
+func (m *Dense) checkOverlap(a blas64.General) bool {
+	mat := m.RawMatrix()
+	if cap(mat.Data) == 0 || cap(a.Data) == 0 {
+		return false
+	}
+
+	off := offset(mat.Data[:1], a.Data[:1])
```
Comment (btracey) [inline mat64/shadow.go]: Why the subslicing?

File: mat64/shadow.go (hunk 9 above)
Comment (kortschak) [inline mat64/shadow.go]: No worries. When I go through and add vector shadowing in earnest I'll add checks for RawMatrixer overlap.

File: mat64/shadow.go (hunk 11 above)
Comment (kortschak) [inline mat64/shadow.go]: We may have been handed something that is zero lengthed.

File: mat64/shadow.go (hunk 10 above)
Comment (kortschak) [inline mat64/shadow.go]: Sure.

Comment (kortschak) [inline mat64/shadow.go]: Is "...offset by off elements after..." OK?

Comment (btracey) [inline mat64/shadow.go]: yes.

File: mat64/shadow.go (hunk 9 above)
Comment (btracey) [inline mat64/shadow.go]: sgtm

Comment (kortschak) [inline mat64/shadow.go]: I've added a TODO to the docs to indicate this intention.

**Comment by kortschak** [comment]
PTAL

File: mat64/doc.go [hunk 12]
```diff
@@ -112,7 +112,34 @@
...
-// detect arbitrary overlap, and so care should be taken when using matrix views.
+// though in many cases this will cause an allocation (see Element Aliasing).
 // An exception to this rule is Copy, which does not allow a.Copy(a.T()).
+//
+// Element Aliasing
+//
+// Most methods in mat64 modify receiver data. It is forbidden for the modified
+// data region of the receiver to overlap the used data area of the input
+// arguments. The exception to this rule is when the method receiver is equal to one
+// of the input arguments, as in the a.Pow(a, 6) call above, or its implicit transpose.
```
Comment (btracey) [inline mat64/doc.go]: I would add something like "mat64 attempts to check for data overlap, for example as constructed with View", possibly in the paragraph saying "don't be clever". Basically, we want to say "mat64 does checking to help you not make mistakes. There are ways to make mistakes using our API, and we try and detect and complain about those. There are lots of ways to create data overlap by circumventing the API. If you need to read this section to see if your overlap is okay, you are being clever. Don't be clever. If you need ... lapack64 ... directly".

In short, I think the current documentation doesn't do enough to explain the goal before telling people not to be clever to get around that goal.

File: mat64/doc.go [hunk 13]
```diff
@@ -112,7 +112,34 @@
...
+//
+// If you need to read this section to understand the behavior of your program,
+// you are being clever. Don't be clever. If you must be clever, blas64 and
+// lapack64 may be used to call the behavior directly.
+//
+// When data is shared between the receiver and input, mat64 will attempt to detect
```
Comment (btracey) [inline mat64/doc.go]: My recommendation for this documentation is:

"mat64 will use the following rules to detect overlap between the receiver and one of the inputs:
 * the input implements one of the Raw methods and
 * the Raw type matches that of the receiver and
 * the address ranges of the backing data slices overlap and
 * either the strides differ or there is an overlap in the used data elements 
If such  an overlap is detected, the method will panic.

The following cases will not panic:
 * the data slices do not overlap 
 * there is pointer identity between the receiver and input values after
    the value has been untransposed if necessary

mat64 does not detect element overlap if the input does not implement a Raw method or if the Raw method differs from that of the receiver. Correct method behavior cannot be guaranteed if there is overlap.

Currently only RawMatrixer aliasing detection is supported."

Comment (kortschak) [inline mat64/doc.go]: >mat64 does not detect element overlap if the input does not implement a Raw method or if the Raw method differs from that of the receiver.

This seems redundant. Is it just there to provide emphasis for the "Correct method behavior cannot be guaranteed if there is overlap" part?

Comment (btracey) [inline mat64/doc.go]: > On Oct 18, 2015, at 6:29 PM, Dan Kortschak <notifications@github.com> wrote:
> 
> In mat64/doc.go <https://github.com/gonum/matrix/pull/263#discussion_r42329418>:
> 
> >  // An exception to this rule is Copy, which does not allow a.Copy(a.T()).
> > +//
> > +// Element Aliasing
> > +//
> > +// Most methods in mat64 modify receiver data. It is forbidden for the modified
> > +// data region of the receiver to overlap the used data area of the input
> > +// arguments. The exception to this rule is when the method receiver is equal to one
> > +// of the input arguments, as in the a.Pow(a, 6) call above, or its implicit transpose.
> > +//
> > +// If you need to read this section to understand the behavior of your program,
> > +// you are being clever. Don't be clever. If you must be clever, blas64 and
> > +// lapack64 may be used to call the behavior directly.
> > +//
> > +// When data is shared between the receiver and input, mat64 will attempt to detect
> mat64 does not detect element overlap if the input does not implement a Raw method or if the Raw method differs from that of the receiver.
> 
> This seems redundant. Is it just there to provide emphasis for the "Correct method behavior cannot be guaranteed if there is overlap" part?
> 
> 

Yes, basically. The idea is “We know you’re wrong if X. We guarantee you’re fine if Y. You’re wrong but we don’t tell you if Z"

Comment (kortschak) [inline mat64/doc.go]: I've modified that section slightly (your words were nicely concise for the rest) to make it really clear what we mean, and to leave open the addition of Vector/Dense checking.

File: mat64/doc.go [hunk 14]
```diff
@@ -112,7 +112,47 @@
...
 // An exception to this rule is Copy, which does not allow a.Copy(a.T()).
+//
+// Element Aliasing
+//
+// Most methods in mat64 modify receiver data. It is forbidden for the modified
+// data region of the receiver to overlap the used data area of the input
+// arguments. The exception to this rule is when the method receiver is equal to one
+// of the input arguments, as in the a.Pow(a, 6) call above, or its implicit transpose.
+//
+// This prohibition is to help avoid subtle mistakes resulting from input corruption
```
Comment (btracey) [inline mat64/doc.go]: 'subtle mistakes when the method needs to read from and write to the same data region. There are many ways ...'

**Comment by btracey** [comment]
Nice documentation wording.

LGTM with the one minor documentation change.

File: mat64/doc.go (hunk 14 above)
Comment (kortschak) [inline mat64/doc.go]: Nice.

**Comment by kortschak** [comment]
Thank you. Very good review.

**Comment by btracey** [comment]
Thank you for putting this together!
